### PR TITLE
[sf-move-subscriptions-api]  business logic for moving subscriptions and unit tests

### PR DESCRIPTION
**Pull request by mkuzdowicz**
## Overview
[sf-move-subscriptions-api]  business logic for moving subscriptions and unit tests

## What is the business logic
- get accessToken from ZUORA
- get account id by Subscription
- update in ZUORA account
-- `CRM Account ID` to `SF Account Id` from request body
-- `sfContactId` to `SF Full contact Id` from request body

## Tests
- [x] CODE
- [x] locally

### CODE REVIEW

File: handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala [hunk 1]
```diff
@@ -2,16 +2,60 @@ package com.gu.sf.move.subscriptions.api
...
+  def message: String
+}
+
+case class FetchZuoraAccessTokenError(message: String) extends MoveSubscriptionServiceError
+
+case class FetchZuoraSubscriptionError(message: String) extends MoveSubscriptionServiceError
+
+case class UpdateZuoraAccountError(message: String) extends MoveSubscriptionServiceError
+
+final case class MoveSubscriptionServiceError2(message: String)
```
Comment (arausuy) [inline handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: Is  there a better name for this?

File: handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala [hunk 2]
```diff
@@ -2,16 +2,60 @@ package com.gu.sf.move.subscriptions.api
...
+  def moveSubscription(moveSubscriptionData: MoveSubscriptionReqBody): EitherT[F, MoveSubscriptionServiceError, MoveSubscriptionServiceSuccess] = {
+    import moveSubscriptionData._
+    logger.info(s"attempt to move $zuoraSubscriptionId subscription to " +
+      s"(Account Id=$sfAccountId ,Full contact Id=$sfFullContactId) SalesForce Contact")
+
+    val moveSubCommand = ZuoraAccountMoveSubscriptionCommand(
+      crmId = sfAccountId,
+      sfContactId__c = sfFullContactId
+    )
+    import Zuora.{accessTokenGetResponseV2, subscriptionGetResponse, updateAccountByMovingSubscription}
```
Comment (arausuy) [inline handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: its a bit confusing having the imports mid function

File: handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala (hunk 1 above)
Comment (mkuzdowicz) [reply handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: uu that is a typo :)

thanks

File: handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala (hunk 2 above)
Comment (mkuzdowicz) [reply handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: i can move them up if you prefer
It depends what style we are used to :)

File: handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala (hunk 1 above)
Comment (mkuzdowicz) [reply handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: actually that class was not used it was a leftover, i deleted it

File: handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala (hunk 2 above)
Comment (mkuzdowicz) [reply handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: moved that upwards

File: handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala [hunk 3]
```diff
@@ -2,16 +2,60 @@ package com.gu.sf.move.subscriptions.api
...
+    (for {
+      accessToken <- accessTokenGetResponseV2(ZuoraConfig, backend)
+        .leftMap(err => FetchZuoraAccessTokenError(err.reason))
+      subscription <- subscriptionGetResponse(ZuoraConfig, accessToken, backend)(SubscriptionName(zuoraSubscriptionId))
+        .leftMap(err => FetchZuoraSubscriptionError(err.reason))
+      updateRes <- updateAccountByMovingSubscription(ZuoraConfig, accessToken, backend)(subscription, moveSubCommand)
+        .leftMap(err => UpdateZuoraAccountError(err.reason))
+    } yield updateRes)
+      .toEitherT[F]
+      .bimap(identity, res => MoveSubscriptionServiceSuccess(res.toString))
```
Comment (frj) [inline handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: is this just map?

Comment (mkuzdowicz) [reply handlers/sf-move-subscriptions-api/src/main/scala/com/gu/sf/move/subscriptions/api/SFMoveSubscriptionsService.scala]: changed to map, thanks for that

**Pull request merged**